File: exercism/scala/allergies/src/main/scala/allergies.scala
object Allergen  {
  sealed trait AllergenBase { val value: Int }

  case object Eggs extends AllergenBase { override val value = 1 }
  case object Peanuts extends AllergenBase { override val value = 2 }
  case object Shellfish extends AllergenBase { override val value = 4 }
  case object Strawberries extends AllergenBase { override val value = 8 }
  case object Tomatoes extends AllergenBase { override val value = 16 }
  case object Chocolate extends AllergenBase { override val value = 32 }
  case object Pollen extends AllergenBase { override val value = 64 }
  case object Cats extends AllergenBase { override val value = 128 }

  val allengens = Seq(Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats)
}

case class Allergies() {
  import Allergen._

  def isAllergicTo(allerge: AllergenBase, n: Int): Boolean = allerge.value <= n

  def allergies(n: Int): Seq[AllergenBase] = {
    def find(remain: Int, ls: Seq[AllergenBase], acc: Set[AllergenBase] = Set.empty): Seq[AllergenBase] = ls match {
      case Nil => acc.toList.sortBy(_.value)
      case hd :: tl =>
        if (isAllergicTo(hd, remain)) find(remain - hd.value, ls, acc + hd)
        else find(remain, tl, acc)
    }

    find(n, allengens.reverse)
  }
}

File: exercism/haskell/allergies/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

import Data.Bits

data Allergen = Cats
              | Chocolate
              | Eggs
              | Peanuts
              | Pollen
              | Shellfish
              | Strawberries
              | Tomatoes deriving (Eq)

value :: Allergen -> Int
value Eggs = 1
value Peanuts = 2
value Shellfish = 4
value Strawberries = 8
value Tomatoes = 16
value Chocolate = 32
value Pollen = 64
value Cats = 128

bitAnd :: Int -> Int -> Bool
bitAnd n m = (.&.) n m == m

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo alg n = bitAnd n $ value alg

allergies' :: [Allergen]
allergies' = [Cats, Chocolate, Eggs, Peanuts, Pollen, Shellfish, Strawberries, Tomatoes]

allergies :: Int -> [Allergen]
allergies n = filter (\a -> isAllergicTo a n) allergies'
